Add more constraint tests

blegat opened the pull request:
Due to the use of `Val{false}`, `aff` can now be a variable or a number in range constraints hence we also need to define `coeftype` for these.

mlubin commented on src/macros.jl:
```diff
@@ -309,18 +304,20 @@ end
 # end
 
 # three-argument constructconstraint! is used for two-sided constraints.
+constructconstraint!(v::Variable, lb::Real, ub::Real) = SingleVariableConstraint(v, MOI.Interval(lb, ub))
```
What happens when `lb` and `ub` are integers, and the solver only supports `Interval{Float64}`?

blegat replied:
Currently this doesn't happen because they are initialized as floating point number
https://github.com/JuliaOpt/JuMP.jl/blob/e6c9f401b63afa630a0af720b6ed507bc0b97dc9/src/macros.jl#L434-L437
I'd say that it is the responsability of the caller of `constructconstraint!` to use relevant types for `lb` and `ub`.

mlubin commented on src/variables.jl:
```diff
@@ -11,6 +11,7 @@ struct Variable <: AbstractJuMPScalar
     m::Model
     index::MOIVAR
 end
+coeftype(::Variable) = Float64
```
This is a sketchy helper function for the macros. I think it should be defined in `macros.jl`.

mlubin commented on src/macros.jl:
```diff
@@ -448,12 +445,12 @@ macro constraint(args...)
                 try
```
This whole `coeftype` thing doesn't seem well thought-out. Maybe we should get rid of it.

blegat commented:
I have replaced the `lbval`/`ubval` error with an error at the `constructconstraint!` level and added the `_error` function to the `constructconstraint!` function to have clearer errors (it also improve the error for range quadratic constraints).
The new error is now
```julia
julia> using JuMP
julia> m = Model()
julia> @variable(m, x)
julia> @variable(m, y)
julia> t = 10.0
julia> @constraint(m, x <= t <= y)
ERROR: In @constraint(m,x <= t <= y): Expected x to be a number.
Stacktrace:
 [1] constraint_error(::Array{Any,1}, ::String) at /home/blegat/.julia/v0.6/JuMP/src/macros.jl:323
 [2] constructconstraint!(::JuMP.#_error#78, ::Float64, ::JuMP.GenericAffExpr{Float64,JuMP.Variable}, ::JuMP.GenericAffExpr{Float64,JuMP.Variable}) at /home/blegat/.julia/v0.6/JuMP/src/macros.jl:319
 [3] macro expansion at ./REPL.jl:97 [inlined]
 [4] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
```

mlubin commented on src/macros.jl:
```diff
@@ -265,25 +265,20 @@ function sense_to_set(sense::Symbol)
-
-constructconstraint!(x::AbstractArray, set::MOI.AbstractScalarSet) = error("Unexpected vector in scalar constraint. Did you mean to use the dot comparison operators like .==, .<=, and .>= instead?")
-constructconstraint!(x::Vector{AffExpr}, set::MOI.AbstractVectorSet) = VectorAffExprConstraint(x, set)
+constructconstraint!(_error::Function, x::AbstractArray, set::MOI.AbstractScalarSet) = error("Unexpected vector in scalar constraint. Did you mean to use the dot comparison operators like .==, .<=, and .>= instead?")
```
error on the RHS should be `_error`

mlubin commented on src/macros.jl:
```diff
@@ -309,17 +304,24 @@ end
 end
 
-# constructconstraint!(aff::Variable, lb::Real, ub::Real) = constructconstraint!(convert(AffExpr,v),lb,ub)
+constructconstraint!(_error::Function, q::QuadExpr, lb, ub) = error("Two-sided quadratic constraints not supported. (Try @NLconstraint instead.)")
```
same here

mlubin commented on src/macros.jl:
```diff
@@ -309,17 +304,24 @@ end
 
-constructconstraint!(q::QuadExpr, lb, ub) = error("Two-sided quadratic constraints not supported. (Try @NLconstraint instead.)")
+function constructconstraint!(_error::Function, expr, lb, ub)
+    @assert !(lb isa Number) || !(ub isa Number)
```
What's the point of this assert?

blegat replied:
It could happen that a call is dispatched to this method with `lb` and `ub` that are not `Number` if there is not `constructconstraint!` method defined for the type of `expr`. In that case, no error will be thrown that the `@constraint` will silently do nothing. We want to avoid that

mlubin commented on src/macros.jl:
```diff
@@ -309,17 +304,24 @@ end
+end
 
-constraint_error(args, str) = error("In @constraint($(join(args,","))): ", str)
+coeftype(::Variable) = Float64
```
Is coeftype still needed for anything?
